File: src/main/java/com/github/zoewithabang/util/HttpRequestHelper.java
package com.github.zoewithabang.util;

import com.github.zoewithabang.model.HttpResponse;
import org.slf4j.Logger;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;

import java.net.URL;
import java.util.Map;
import java.util.stream.Collectors;


public class HttpRequestHelper
{
    private static Logger LOGGER = Logging.getLogger();
    
    public static HttpResponse performGetRequest(String urlString, String query, Map<String, String> requestProperties) throws MalformedURLException, IOException
    {
        URL url;
        HttpURLConnection connection;
        
        try
        {
            if(query == null)
            {
                url = new URL(urlString);
            }
            else
            {
                url = new URL(urlString + "?" + query);
            }
        }
        catch(MalformedURLException e)
        {
            if(query == null)
            {
                LOGGER.error("MalformedURLException on creating URL object from string '{}'.", urlString, e);
            }
            else
            {
                LOGGER.error("MalformedURLException on creating URL object from string '{}' with query '{}'.", urlString, query, e);
            }
            throw e;
        }
        
        try
        {
            connection = (HttpURLConnection)url.openConnection();
            for(Map.Entry<String, String> entry : requestProperties.entrySet())
            {
                connection.setRequestProperty(entry.getKey(), entry.getValue());
            }
            
            int responseCode = connection.getResponseCode();
            
            if(200 <= responseCode && responseCode <= 299) //success
            {
                LOGGER.debug("Successful response code {}, getting stream result.", responseCode);
                return new HttpResponse(getInputStreamResultFromConnection(connection), connection.getURL().toString());
            }
            else if(responseCode >= 400) //error
            {
                LOGGER.error("Error response code {}, getting error stream result.", responseCode);
                LOGGER.error(getErrorStreamResultFromConnection(connection));
                return null;
            }
            else
            {
                LOGGER.warn("Unexpected response code from connection: {}", connection);
                return null;
            }
        }
        catch(IOException e)
        {
            LOGGER.error("IOException on performing GET request for URL '{}' and request properties '{}'.", url, requestProperties, e);
            throw e;
        }
    }
    
    private static byte[] getInputStreamResultFromConnection(HttpURLConnection connection) throws IOException
    {
        try(InputStream input = connection.getInputStream();
            ByteArrayOutputStream output = new ByteArrayOutputStream())
        {
            byte[] chunk = new byte[4096];
            int length;
            
            while((length = input.read(chunk, 0, chunk.length)) > 0)
            {
                output.write(chunk, 0, length);
            }
            
            return output.toByteArray();
        }
        catch(IOException e)
        {
            LOGGER.error("IOException on getting input stream result for connection '{}'.", connection, e);
            throw e;
        }
    }
    
    private static String getErrorStreamResultFromConnection(HttpURLConnection connection) throws IOException
    {
        try(BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getErrorStream())))
        {
            return reader.lines()
                .collect(Collectors.joining());
        }
        catch(IOException e)
        {
            LOGGER.error("IOException on getting error stream result for connection '{}'.", connection, e);
            throw e;
        }
    }
}

File: src/main/kotlin/com/github/zoewithabang/util/HttpRequestHelper.kt
package com.github.zoewithabang.util

import com.github.zoewithabang.model.HttpResponse
import org.apache.logging.log4j.LogManager
import org.slf4j.Logger

import java.io.*
import java.net.HttpURLConnection
import java.net.MalformedURLException

import java.net.URL
import java.util.stream.Collectors


object HttpRequestHelper {
    private val logger = LogManager.getLogger("HttpRequestHelper")

    @Throws(MalformedURLException::class, IOException::class)
    fun performGetRequest(urlString: String, query: String?, requestProperties: Map<String, String>): HttpResponse? {
        val url: URL
        val connection: HttpURLConnection

        try {
            url = if (query == null) {
                URL(urlString)
            } else {
                URL("$urlString?$query")
            }
        } catch (e: MalformedURLException) {
            if (query == null) {
                logger.error("MalformedURLException on creating URL object from string '{}'.", urlString, e)
            } else {
                logger.error(
                    "MalformedURLException on creating URL object from string '{}' with query '{}'.",
                    urlString,
                    query,
                    e
                )
            }

            throw e
        }

        try {
            connection = url.openConnection() as HttpURLConnection
            for (entry in requestProperties.entries) {
                connection.setRequestProperty(entry.key, entry.value)
            }

            val responseCode = connection.responseCode

            return when {
                responseCode in 200..299 -> {
                    logger.debug("Successful response code {}, getting stream result.", responseCode)

                    HttpResponse(getInputStreamResultFromConnection(connection), connection.url.toString())
                }
                responseCode >= 400 -> {
                    logger.error("Error response code {}, getting error stream result.", responseCode)
                    logger.error(getErrorStreamResultFromConnection(connection))

                    null
                }
                else -> {
                    logger.warn("Unexpected response code from connection: {}", connection)

                    null
                }
            }
        } catch (e: IOException) {
            logger.error(
                "IOException on performing GET request for URL '{}' and request properties '{}'.",
                url,
                requestProperties,
                e
            )

            throw e
        }

    }

    @Throws(IOException::class)
    private fun getInputStreamResultFromConnection(connection: HttpURLConnection): ByteArray {
        try {
            connection.inputStream.use { input ->
                ByteArrayOutputStream().use { output ->
                    val chunk = ByteArray(4096)
                    var length = input.read(chunk, 0, chunk.size)

                    while (length > 0) {
                        output.write(chunk, 0, length)

                        length = input.read(chunk, 0, chunk.size)
                    }

                    return output.toByteArray()
                }
            }
        } catch (e: IOException) {
            logger.error("IOException on getting input stream result for connection '{}'.", connection, e)

            throw e
        }

    }

    @Throws(IOException::class)
    private fun getErrorStreamResultFromConnection(connection: HttpURLConnection): String {
        try {
            BufferedReader(InputStreamReader(connection.errorStream)).use { reader ->
                return reader.lines()
                    .collect(Collectors.joining())
            }
        } catch (e: IOException) {
            logger.error("IOException on getting error stream result for connection '{}'.", connection, e)

            throw e
        }
    }
}
